src/fges.jl

```diff
@@ -0,0 +1,428 @@
+
+        return new{M, baseType}(data, normAugData, numFeatures, numObservations, penalty)
+    end
+end  
+
+# Simple structure to hold the current edge, a subset of neighbors, and a score change
+Base.@kwdef mutable struct Step{A,B}
+    edge::Edge = Edge(1,2)
+    subset::Vector{A} = A[]
+    Δscore::B = zero(B)
+end
+
+
+####################################################################
+# Base Function Overloads
+####################################################################
+
+# Print method to display the Step
+function show(io::IO, newStep::Step{A,B}) where {A,B}
+    print(io, "Edge: $(newStep.edge), Subset: $(newStep.subset), Δscore: $(newStep.Δscore)")
+end
+
+# The @memoize macro has to check if ParseData is the same argument. 
+# We only ever define one immutable ParseData. We check equality of data.
+==(a::T, b::T) where T <: ParseData = a.data == b.data
+
+
+####################################################################
+# Main Entry point for the Algorithm
+####################################################################
+
+"""
+    fges(data; penalty = 1.0, verbose=false)
+
+Compute a causal graph for the given observed data using GES.
+"""
+function fges(data; penalty = 1.0, verbose=false)
+
+    # data type / precision
+    scoreT = eltype(data)
+
+    # Parse the inputted data
+    dataParsed = ParseData(data, penalty)
+
+    # Create an empty graph with one node for each feature
+    g = DiGraph(dataParsed.numFeatures)
+
+    return fges_internal(g, scoreT, dataParsed; penalty, verbose)
+end
+
+function fges_internal(g, scoreT, data; penalty = 1.0, verbose=false)
+
+    verbose && println("Start forward search")
+    # Perform the forward search 
+    
+    ges_search!(g, Step{Int, scoreT}(), data, Insert!, verbose)
+
+    verbose && println("Start backward search")
+    # Perform the backward search 
+    ges_search!(g, Step{Int, scoreT}(), data, Delete!, verbose)
+    
+    return g # Return the graph
+
+end
+
+####################################################################
+# Insert and Delete Operators
+####################################################################
+"""
+
+    Insert!(g, x, y, T)
+
+Inserts x->y and directs previously undirected edges t->y, t ∈ T.
+Here x and y are not adjacent and T are undirected-neighbors of y 
+that are not adjacent to x.
+"""
+function Insert!(g, x, y, T)
+    add_edge!(g, x → y)
+    
+    # Orient all edges in T incident into child node
+    for t ∈ T
+        orientedge!(g, t → y)
+    end
+    return g
+end
+
+
+
+"""
+
+    Deletes!(g, x, y, H)
+
+Deletes x-y or x->y and directs previously undirected edges x->h and y->h
+for h in H.
+"""
+function Delete!(g, x, y, H)
+
+    # Remove directed or undirected edges (x→y and x-y)
+    rem_edge!(g, x → y)
+    rem_edge!(g, y → x)
+    
+    # Orient all vertices in H toward x and y
+    for h ∈ H
+        if has_both(g, x, h) # reading literally Definition 13, Chickering
```

Comment:
This is also tricky!

Reply:
Did you figure it all out? Or are there some ambiguities left?